File: libs/power-log-parser/src/lib/parsers/entity-chosen-parser.ts
import { CardIds, GameTag } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventProvider } from '../game-event';
import { Choice, Node, NodeType } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class EntityChosenParser implements ActionParser {
	readonly ParserName = 'EntityChosenParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(node: Node, _stateType: StateType): boolean {
		return node.Type === NodeType.Choice && this.ParserState.CurrentChosenEntites != null;
	}

	AppliesOnCloseNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const choice = node.Object as Choice;
		const ptlState = this.StateFacade.PtlState?.GameState;
		const keyInGS = this.GameState.CurrentEntities.has(choice.Entity);
		const chosenEntity =
			ptlState?.CurrentEntities?.has(choice.Entity)
				? ptlState.CurrentEntities.get(choice.Entity)!
				: keyInGS
					? this.GameState.CurrentEntities.get(choice.Entity)!
					: null;
		if (chosenEntity == null) {
			return null;
		}

		const isCopy = ptlState?.CurrentEntities?.has(chosenEntity.GetTag(GameTag.LINKED_ENTITY)) ?? false;
		const originalEntity = isCopy
			? ptlState!.CurrentEntities.get(chosenEntity.GetTag(GameTag.LINKED_ENTITY))!
			: null;
		const controllerId = chosenEntity.GetEffectiveController();

		return [
			GameEventProvider.Create(
				choice.TimeStamp,
				'ENTITY_CHOSEN',
				() => {
					const gsChosenEntity = this.GameState.CurrentEntities.get(choice.Entity);
					let creatorEntityId = gsChosenEntity?.GetTag(GameTag.CREATOR) ?? -1;
					if (creatorEntityId === -1) {
						creatorEntityId = gsChosenEntity?.GetTag(GameTag.DISPLAYED_CREATOR) ?? -1;
					}
					const creatorEntity = ptlState?.CurrentEntities?.has(creatorEntityId)
						? ptlState!.CurrentEntities.get(creatorEntityId)!
						: null;
					const creatorCardId = creatorEntity?.CardId;

					if (
						creatorCardId === CardIds.SuspiciousPirate ||
						creatorCardId === CardIds.SuspiciousAlchemist ||
						creatorCardId === CardIds.SuspiciousUsher ||
						creatorCardId === CardIds.SuspiciousPeddler
					) {
						creatorEntity!.KnownEntityIds.push(chosenEntity.Id);
					}

					return {
						Type: 'ENTITY_CHOSEN',
						Value: {
							CardId: chosenEntity.CardId,
							ControllerId: controllerId,
							LocalPlayer: this.StateFacade.LocalPlayer,
							OpponentPlayer: this.StateFacade.OpponentPlayer,
							EntityId: chosenEntity.Id,
							AdditionalProps: {
								OriginalEntityId: originalEntity?.Id,
								Context: {
									CreatorEntityId: creatorEntityId,
									CreatorCardId: creatorCardId,
								},
							},
						},
					};
				},
				true,
				node,
				null,
				400,
			),
		];
	}

	CreateGameEventProviderFromClose(_node: Node): GameEventProvider[] | null {
		return null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/EntityChosenParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using HearthstoneReplays.Parser.ReplayData.Meta;
using System.Linq;

namespace HearthstoneReplays.Events.Parsers
{
    public class EntityChosenParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public EntityChosenParser(ParserState ParserState, StateFacade helper)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = helper;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return node.Type == typeof(Choice)
                && ParserState.CurrentChosenEntites != null;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return false;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var choice = node.Object as Choice;
            // Why use Ptl here? 
            var ptlState = StateFacade.PtlState.GameState;
            var keyInGS = GameState.CurrentEntities.ContainsKey(choice.Entity);
            // Special case for Sphere of Sapience, as the corresponding block is not properly closed
            FullEntity chosenEntity = (ptlState?.CurrentEntities?.ContainsKey(choice.Entity) ?? false)
                ? ptlState.CurrentEntities[choice.Entity]
                : keyInGS
                    ? GameState.CurrentEntities[choice.Entity]
                    : null;
            if (chosenEntity == null)
            {
                return null;
            }

            // Entities offered in chocies are often côpies
            var isCopy = ptlState.CurrentEntities.ContainsKey(chosenEntity.GetTag(GameTag.LINKED_ENTITY));
            var originalEntity = isCopy ? ptlState.CurrentEntities[chosenEntity.GetTag(GameTag.LINKED_ENTITY)] : null;
            var controllerId = chosenEntity.GetEffectiveController();

            return new List<GameEventProvider> { GameEventProvider.Create(
                choice.TimeStamp,
                "ENTITY_CHOSEN",
                () =>
                {
                    var gsChosenEntity = GameState.CurrentEntities.GetValueOrDefault(choice.Entity);
                    // Because when the opponent Discovers, the CREATOR tag is set right AFTER the discover, so we need to wait a little bit
                    var creatorEntityId = gsChosenEntity?.GetTag(GameTag.CREATOR) ?? -1;
                    // When opponent discovers
                    if (creatorEntityId == -1)
                    {
                        creatorEntityId = gsChosenEntity?.GetTag(GameTag.DISPLAYED_CREATOR) ?? -1;
                    }
                    var creatorEntity = ptlState.CurrentEntities.ContainsKey(creatorEntityId) ? ptlState.CurrentEntities[creatorEntityId] : null;
                    var creatorCardId = creatorEntity?.CardId;

                    if (creatorCardId == CardIds.SuspiciousPirate
                        || creatorCardId == CardIds.SuspiciousAlchemist
                        || creatorCardId == CardIds.SuspiciousUsher
                        || creatorCardId == CardIds.SuspiciousPeddler)
                    {
                        creatorEntity.KnownEntityIds.Add(chosenEntity.Id);
                    }

                    return new GameEvent
                    {
                        Type = "ENTITY_CHOSEN",
                        Value = new
                        {
                            CardId = chosenEntity.CardId,
                            ControllerId = controllerId,
                            LocalPlayer = StateFacade.LocalPlayer,
                            OpponentPlayer = StateFacade.OpponentPlayer,
                            EntityId = chosenEntity.Id,
                            AdditionalProps =  new {
                                OriginalEntityId = originalEntity?.Id,
                                Context = new
                                {
                                    CreatorEntityId = creatorEntityId,
                                    CreatorCardId = creatorCardId,
                                }
                            }
                        }
                    };
                },
                true,
                node, 
                null,
                waitFor: 400) };
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return null;
        }
    }
}